rfcs/0079-tvmscript-metaprogramming.md

```diff
@@ -0,0 +1,398 @@
+IRBuilder to let user construct IR imperatively. IRBuilder will be used by the
+parser, as well as by users directly as described in the feature F4. IRBuilder
+allows user to write code in a style that’s similar to TVMScript, while it’s
+being executed as host program. For example, 
+
+```python
+from tvm.script.builder import Builder, def_, def_many
+from tvm.script import tir as T
+
+with Builder() as b:
+  with T.prim_func():
+    T.func_name("main")
+    buffer_a = T.Buffer((128, 128, 128), "float32")
+    buffer_b = T.Buffer((128, 128, 128), "float32")
+    arg_a = T.arg("A", buffer_a)
+    arg_b = T.arg("B", buffer_b)
+    with T.grid(128, 128, 128) as (i, j, k):
+      def_many(["i", "j", "k"], [i, j, k])
+      with T.block(name="block"):
+        vi = def_("vi", T.axis.spatial(128, i))
+        vj = def_("vj", T.axis.spatial(128, j))
+        vk = def_("vk", T.axis.reduce(128, k))
+
+f = b.get() # f is a PrimFunc
+```
+
+produces similar result to 
+
+```python
+@T.prim_func
+def main(
+  A: T.Buffer(shape=(128, 128, 128), dtype="float32"),
+  B: T.Buffer(shape=(128, 128, 128), dtype="float32"),
+) -> None:
+  for i, j, k in T.grid(128, 128, 128):
+      with T.block("block"):
+          vi = T.axis.S(128, i)
+          vj = T.axis.S(128, j)
+          vk = T.axis.R(128, k)
+```
+
+The implementation of IRBuilder will be in C++ so that it can be used in an
+environment without Python. Python binding will be created to expose IRBuilder
+to TVMScript parser.
+
+## Parse-time evaluation
+
+To support metaprogramming, the TVMScript parser needs to evaluate the
+expression using Python interpreter. All metaprogramming
+features we discussed above can be implemented through this parse-time
+evaluation. Using the same `gen_matmul` example from F1,
+
+```python
+def gen_matmul(n, m) -> None:
+  @T.prim_func
+  def f(A: T.Buffer[(n, m)]):
+    ...
+  return f
+```
+
+What parser does here is to:
+
+1. Collect the environment inside `gen_matmul`, getting a dictionary
+    1. All primitive types will be captured automatically, while advanced types
+       (like function) needs to be explicitly declared in the decorator to be
+       captured (for example, `@T.prim_func(capture=[get_box_coordinates])`)
+2. Call the corresponding function from IRBuilder, as the parser visits the AST
+   of function `f`.
+    1. When visiting the function argument, call `eval` on its type annotation, 
+       with the environment captured in the first step.
+    2. `T.Buffer[(n, m)]` gets evaluated to a value with type `tir.Buffer`.
+    3. Call the IRBuilder API `T.arg("A", buffer)` to add an arg to the function
+       that’s being constructed
+
+Another example,
+
+```python
+for *i in T.grid(*A.shape):
+  ...
+```
+
+The parser will:
+
+1. Evaluate the expression `T.grid(*A.shape)` by the step described above.
+   `T.grid` returns a value that is nearly equivalent to `List[Var]`. 
+2. Call `exec` on a specially constructed statement `*i = __tvm_rhs_var__`,
```

Comment:
This is a good question. The original RFC indeed lacks some discussion on the necessity and security implication behind the usage of `eval` and `exec`. I updated the RFC to include a section for it.

To answer this comment directly, the left hand side of assignment can get very complicated, like
```
a, (b, *c, _), d = rhs
```
It's much easier to use `exec`, compared to writing our own small Python interpreter. Since the tokens on the LHS comes from user's code without modification, and the auto capturing behavior is limited, the usage of `exec` here does not create additional security risk for user.

> using exec means folks would need to essentially strip this code out from places that were productionizing TVM

I think it's a little bit far-fetched. I agree that the usage of `eval` and `exec` needs strong rationale and careful security review. Also using them for untrusted input is a big security hole. But there are still legitimate use cases. In our case the input comes from user's code directly, which should be considered safe. There is a similar usage of `exec` in `torch.jit` https://github.com/pytorch/pytorch/blob/master/torch/jit/frontend.py#L465 (constructing a function from template with user's input).

Reply:
This is an interesting discussion and both sides seem to have valid points - `eval` and `exec` can be very useful but they are susceptible to misuse. If they are irreplaceable, can we provide more guideline (e.g., good/bad use-cases) to avoid the misuses?

Reply:
It will be quite challenging to have a comprehensive guideline over the usage of `eval` and `exec`. In our particular case, the guideline is to always limit the input of `eval` and `exec` to fragments of user's code (or the equivalent snippet, like `<lhs> = __tvm_rhs_var__`). By doing this, we eliminate the possibility that unknown code is executed without user's awareness.

Another effect of using `eval` is that the timing of the evaluation changes, an extreme example would be (although not something people will write in TVMScript) :
```
@T.prim_func
def f():
    ...
    if False:
        nuke_home_directory()
    ...
```
User would expect:
1. Nothing from `f` will run if `f` is not called by other code
2. `nuke_home_directory` will not be called because it's dead code

To minimize the surprise to users, we ask user to explicitly capture those function before they can be evaluated in the parser, to make it clear that the timing of evaluation is different than ordinary Python code.